Adding Surface Magnetism Solver + Hand Ray Example scene

cre8ivepark opened the pull request:
**NOTE:** Dependent on #6304 PR

![MRTK_SurfaceMagExample](https://user-images.githubusercontent.com/13754172/66724623-1c4eea80-eddd-11e9-986b-91b92bcdd4cd.gif)
## Overview
Currently, there is no easy way to make an object attached to the endpoint of the Hand Ray and make it placed on a surface with the surface normal aligned. Adding a new example scene for the Spatial Awareness + Surface Magnetism

## Changes
In SurfaceMagnetism.cs, I added Hand Ray option. However, **I am not sure if this is the right place to add since the section name is 'Raycasting Direction Mode'. I feel this could go into SolverHandler's Tracked Target Object. Need help here.** 

                    case RaycastDirectionMode.HandRay:
                        PointerUtils.TryGetPointerEndpoint<IMixedRealityPointer>(Handedness.Both, InputSourceType.Hand, out endPoint);
                        break;

![2019-10-13 12_51_48-Window](https://user-images.githubusercontent.com/13754172/66724628-2bce3380-eddd-11e9-8370-a1f8a3e15b4d.png)
![2019-10-13 13_04_49-Window](https://user-images.githubusercontent.com/13754172/66724634-44d6e480-eddd-11e9-9f18-85f61756e813.png)

## Example Scene
Added a new example scene **SurfaceMagnetismSpatialAwarenessExample** to demonstrate a slate following the Spatial Mapping surface. 
![2019-10-13 17_36_30-Unity 2018 4 7f1 Personal - SurfaceMagnetismSpatialAwarenessExample unity - MRTK](https://user-images.githubusercontent.com/13754172/66724886-127ab680-ede0-11e9-9bdd-31d1d7f8f838.png)


- Fixes: #6279

cre8ivepark commented:
@Troy-Ferrell the Hand joint options make the object attach to the hand joints, not to the endpoint of the ray. I couldn't find a wat to achieve it. However, for 'Head' option, it already gives me the capability to attach object to the endpoint of the gaze, combined with Surface Magnetism's Raycast Direction

Troy-Ferrell commented:
> Hand

When this property is set as such on Surface Magnetism, it will use the forward ray direction of the tracked target type on the associated SolverHandler.
Current Raycast Direction Mode = Tracked Target Forward

In the case of head, it is the camera's forward gaze.
In the case of the hand, it will be the forward vector of the palm (or whatever joint selected) transform. 

So it actually should work. I swear I was using it right before GA. 

I'll look into this tomorrow but the current API should allow what you are trying to do @cre8ivepark

cre8ivepark commented:
@Troy-Ferrell That would be amazing! I'll also check tonight with palm + tracked target forward option.

Troy-Ferrell commented:
> @Troy-Ferrell That would be amazing! I'll also check tonight with palm + tracked target forward option.

Actually I already checked. Sorry, that is what I meant by things were currently broken and why I updated your bug title. For some reason, the forward vector on the hand palm join (NOTE at least in input simulation) was not working with the Surface Magnetism (TrackedTargetForward option).

I'm gonna try to debug later today

cre8ivepark commented:
One more question: will Palm + forward raycast give us the same result as Hand Ray's endpoint? Because the Hand Ray's ray is coming from the shoulder, the target object position might not perfectly match the endpoint of the hand ray?

Troy-Ferrell commented:
> One more question: will Palm + forward raycast give us the same result as Hand Ray's endpoint? Because the Hand Ray's ray is coming from the shoulder, the target object position might not perfectly match the endpoint of the hand ray?

I figured out why hand joint was failing. But you bring up a good point. It wouldn't exactly be the same hand ray as used by the pointers. 

I think I have a solution though that should fit nicely into the existing framework to fix the bug and be a bit cleaner code than what is in SolverHandler. I'll create a PR for that soon.

@cre8ivepark , we should still keep this PR though. You did a fantastic job helping with the Interactable scene update. It would be incredible if you could apply your same magic to the Solver example scene.

I'll sync with you soon to talk about it.

![SM_Fix](https://user-images.githubusercontent.com/25975362/66880772-30c7ea00-ef79-11e9-82d9-482db7068378.gif)

cre8ivepark commented:
@Troy-Ferrell Looks amazing! Once you open the PR, I'll add the Spatial Awareness + Surface Magnetism example scene.

Troy-Ferrell commented:
@cre8ivepark, pull or wait for this branch and then update your PR
#6304

cre8ivepark commented:
@Troy-Ferrell 
Reverted SurfaceMagnetism.cs
Updated the scene example to use 'Controller Ray' as a tracked target type.
Created new profiles for the example scene. (to enable Spatial Awareness)
Tested on HoloLens 2 device.
<img width="500" alt="2019-10-28 13_12_27-MixedRealityToolkit-Unity - Microsoft Visual Studio" src="https://user-images.githubusercontent.com/13754172/67714275-cf464900-f984-11e9-9f22-c1aa77d607b1.png">

keveleigh commented:
/azp run

Troy-Ferrell commented on Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleMRTKConfigurationProfile.asset:
```diff
@@ -0,0 +1,54 @@
+--- !u!114 &11400000
+MonoBehaviour:
+  m_ObjectHideFlags: 0
+  m_CorrespondingSourceObject: {fileID: 0}
```
Is there anything special here other than turning on SA? If not, I'd say let's just re-use "DefaultMixedRealityToolkitConfigurationProfile".

cre8ivepark replied:
In this profile, I made the mesh invisible to reduce visual noise.

Troy-Ferrell replied:
I'd say use the default and then add hand menu or toggle menu to switch out between occluded vs visible states in SA system

cre8ivepark replied:
Great idea!

julenka commented on Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleSpatialAwarenessSystemProfile.asset:
```diff
@@ -0,0 +1,23 @@
+  m_GameObject: {fileID: 0}
+  m_Enabled: 1
+  m_EditorHideFlags: 0
+  m_Script: {fileID: 11500000, guid: 19f279aded72cb741b4de89a54359dd4, type: 3}
```
Will you be removing this asset, per Troy's comment?

cre8ivepark replied:
Yes. And adding toggle button for spatial mesh visual.

cre8ivepark commented:
Doing some final device testings for
- Spatial Mesh visual toggle button
- Added another plane example object with tracked object 'head' option for HoloLens 1st gen users.

cre8ivepark commented:
Spatial Mesh visual toggle button
![2019-10-31 16_15_37-Unity 2018 4 6f1 Personal - SurfaceMagnetismSpatialAwarenessExample unity - MRTK](https://user-images.githubusercontent.com/13754172/67993182-f018d380-fbfc-11e9-8596-3e0199e0d8ac.png)

Added another plane example object with tracked object 'head' option for HoloLens 1st gen users.
![2019-10-31 16_15_26-Unity 2018 4 6f1 Personal - SurfaceMagnetismSpatialAwarenessExample unity - MRTK](https://user-images.githubusercontent.com/13754172/67993184-f3ac5a80-fbfc-11e9-9fae-0b7217211584.png)

Deleted custom profile for the spatial awareness

keveleigh commented:
/azp run
